common/iterator/sorted/IntersectForwardableIterator.java

```diff
@@ -21,40 +21,52 @@
 import com.vaticle.typedb.core.common.exception.TypeDBException;
 import com.vaticle.typedb.core.common.iterator.FunctionalIterator;
 
-import java.util.Iterator;
+import java.util.HashSet;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.NoSuchElementException;
+import java.util.Set;
 import java.util.function.Function;
 import java.util.function.Predicate;
 
 import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
 
+/**
+ * Intersection iterator, defined as intersection based on the comparator only, not the equality function.
+ * If the equality and comparator functions are one-to-one, the intersection is effectively an equality-based intersection.
+ *
+ * The intersection of iterators with multiple values that are comparably == 0 is defined as the distinct
+ * union of all of these elements.
+ */
 public class IntersectForwardableIterator<T extends Comparable<? super T>, ORDER extends SortedIterator.Order>
         extends AbstractSortedIterator<T, ORDER>
         implements SortedIterator.Forwardable<T, ORDER> {
 
     private final List<Forwardable<T, ORDER>> iterators;
 
     private T candidate;
-    private int candidateSource;
+    private Forwardable<T, ORDER> candidateSource;
+    private final Set<T> intersectionValues;
+    private final List<Forwardable<T, ORDER>> intersectionIterators;
```

Comment:
keep track of iterators that are intersect at the last candidate, and the set of values they have produced.
`intersectionValues` ensures we do not return duplicates from the intersection.